### ISSUE TITLE
omit_local_variable_types has spurious diagnostics when the variable type impacts inference

### ISSUE THREAD

**Issue by natebosch** [dart-lang/linter#4772]
```dart
void main() async {
  final String s1 = await someGeneric(['']); // omit_local_variable_types
  print(typeOf(s1)); // String
  // Removing the annotation results in a static type change:
  final s2 = await someGeneric(['']);
  print(typeOf(s2)); // dynamic
}

Future<T> someGeneric<T, S extends List<T>>(S v) async => v.first;

Type typeOf<T>(T arg) => T;
```

The lint fires and asks for the local variable type to be removed, but doing so changes the static type of the variable. We don't correctly check when the variable impacts inference, because we only check whether the return type _is_ a type variable.

https://github.com/dart-lang/sdk/blob/6885e82cda23c253fbe5e8720d334070ebda9d86/pkg/linter/lib/src/rules/omit_local_variable_types.dart#L132

Once the type is more complicated, like `Future<T>`, the lint misfires.

cc @pq @yanok

**Comment by srawlins** [comment]
Sigh, yeah I think we have to file this in the bucket of "in order to implement this perfectly correctly, the lint rule needs to run _during_ inference" (or at least have knowledge of context types). ☹️

**Comment by yanok** [comment]
To be _perfectly_ correct, right, lint needs to be able to re-run inference for the expression without a context type. But to just exclude false positives, we could simple modify the check from "is a type variable" to "contains a type variable". This is a conservative check, so there will be false negatives. But at least it will be more consistent than it is now, since now we can get _both_ false negatives and false positives.

**Comment by bwilkerson** [comment]
> ... the lint rule needs to run _during_ inference ...

For what it's worth, I'm not convinced that even that would be enough. I think it's the case that we would need to be able to ask "what if" questions. Things like "what would the (inferred) type of this variable be if we were to remove the type annotation. Yes, that requires running inference over the modified code, but I don't think we'd be able to answer that question while running inference over the unmodified code.

**Comment by srawlins** [comment]
> But to just exclude false positives, we could simple modify the check from "is a type variable" to "contains a type variable".

I think this is very pragmatic! 👍 

> I think it's the case that we would need to be able to ask "what if" questions.

Oh I think you're right. Yeah this might need to include flow analysis of statements and expressions that follow the variable declaration, with a "hypothetical" type. This is the same conundrum as the `unnecessary_cast` false positives.

**Comment by srawlins** [comment]
@pq as a P1, this ticket should be actively worked on. Is this one you are planning to take, or should I assign to someone else?

**Comment by pq** [comment]
Agreed. I'm not actively on this one so if anyone is game, they're welcome to jump on it!

**Comment by eernstg** [comment]
There is a way to simplify the "false positives are complex" issue with `omit_local_variable_types`: If we milden this lint such that it only asks developers to remove a declared type when the initializing expression has an 'obvious' type (as proposed in https://github.com/dart-lang/linter/issues/3480) then we would probably have very few false positives in the first place.

**Comment by pq** [comment]
@stereotype441: more fodder for the flow analysis conversation
